add special handling for resize_() in functionalization pass

bdhirsh opened the pull request:
This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527
* #76358
* __->__ #77714
* #77713
* #77132
* #77716

bdhirsh committed:
add special handling for resize_() in functionalization pass

[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

ezyang commented on aten/src/ATen/FunctionalizeFallbackKernel.cpp:
```diff
@@ -81,6 +85,84 @@ namespace {
+     at::AutoDispatchSkipFunctionalize guard;
+     at::Tensor tmp_output = at::_ops::resize_::call(self_, size, memory_format);
+     return self;
+  }
```
nit: it seems like it would be mildly better to do this before unwrapping?

ezyang reviewed (approved):
I didn't audit the exact mechanics carefully but the high level algorithm LGTM

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "add special handling for resize_() in functionalization pass"

This PR adds support for `resize_()` in functionalization.

I'm actually kinda surprised this didn't surface as a problem earlier - LTC tests that used resize_() happened to not fail because originally, functionalization would treat `resize_()` as a no-op and redispatch (which is correct in the basic case, but totally wrong in more complex cases). With the new codegen PR further down in the stack that adds a new `resize.functional` operator, `resize_()` now does the wrong thing.

I wrote a custom kernel for it, because it require some special handling. In particular:
- If you `resize_()` down to a smaller storage, then `resize_()` acts as an inplace_view operator - you're basically viewing a new, smaller slice of the original tensor. I implemented this logic using `as_strided`
- If you `resize_()` to a larger input, then we need special handling - we need to re-create the underlying storage.

I'm also banning the following usage of `resize_()` in functionalization:
```
a = torch.ones(2)
b = a.view(2)
b.resize_(4) # try resize a **view** tensor to a larger size
```

My reasoning is:
- `resize_()` for the most part is only meant to be used for resizing size=0 tensors, that shouldn't have any outstanding views
- there are some inherent problems with the design of functionalization that makes the above code hard to support.

Why is the above code hard to support? Functionalization basically assumes that:
- every view tensor comes from a "base", that corresponds to "all of the data"
- `b = view_op(a)` always implies that b is viewing a subset of a (and not the other way around)

These two assumptions are important because of the view-replay logic. We sync mutations across aliases by first propagating a mutation back up to the base, and then replaying the views for each alias from the newly updated base tensor. In the above example, we wouldn't be able to regenerate `b` from `a`, since `b` points to extra storage that `a` doesn't know about.

I suppose that we're also technically banning code like this:
```
a = torch.ones(2)
a_view = a.view(2)
a.resize_(4)
```

Where `a` is technically the "base", but it still has outstanding views. This would be possible to support but still more complicated, so I'm holding off on it for now.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
